### DISCUSSION TITLE
Texels to pixel contribution

### DISCUSSION THREAD

**Discussion by sojiro0** [facebookresearch/pytorch3d#1910] [General]
Hello,
i am trying to implement a basic concept for a very long time but i cannot find any documentation or example so hopefully i can get some help here.
I want to get for all the pixels (in the rendered view) the texels that contributed. So for example if i render an image of size 512² then i want to get a [512,512,T,2] tensor where T will be the texels and 2 their coordinates in the uv space.
I managed to get 512,512,2 so for each pixel i get one texel but i want to get all the texels
A visualization of what i wanna achieve is this: 
![image](https://github.com/user-attachments/assets/97080aaf-eabd-47b4-84f6-6dc967767827)
As seen here, for every pixel i wanna get the quadrilateral in the uv space

Thanks!

**Comment by bottler** [comment]
I think there's a trick to do this.

You can change the map in your `textures_uv` object to just contain coordinates (i.e. to have 2 channels, one of which goes from 0 to 1 in the x direction, the other goes in the y direction, like [this](https://github.com/facebookresearch/pytorch3d/blob/main/pytorch3d/renderer/mesh/textures.py#L1594-L1599) but converted to float32 and the right shape for the map. Then if `fragments` is the rasterizer output, you can run `texels = meshes.sample_textures(fragments)` and the output should be what you want.

**Comment by sojiro0** [comment]
Hi @bottler,
if i understand your suggestion correctly i ll only be able to get one coordinate (i.e. one texel) per pixel. I already can get this by using 
`interpolate_face_attributes(...)`. This returns one texel per pixel. So if i have a 512² rendered image, my code will also create a [1,512,512,2] tensor with uv coordinates.
This is not what i need. I need **all** the texels that contribute **per pixel**.
Another visualization of what i need is this:
![image](https://github.com/user-attachments/assets/e894f1b3-f850-491a-b680-3ce35ed245d5)
Here (as in my previous screenshot) i have one pixel and for this pixel many texels contribute. Lets say in this example that 4 texels contribute. Again in the texture map a quadrilateral shape is formed **per pixel**.
I assume the rasterizer already does this, there should be an easy way of doing this. Maybe getting the coordinates for all the texels that contributed per pixel (a list of lists as pixels will have an uneven number of contributing texels), or the coordinates for the 4 points of the quadrilateral shape per pixel.
Isn't this how the rasterizer works either or else?

**Reply by bottler** [reply]
No that's not right. Blending/shading is the process which combines information from multiple faces into a pixel, and my suggestions operates without that. The `fragments` which the rasterizer outputs have a `faces_per_pixel` dimension, as will the output of `sample_textures`.

**Reply by sojiro0** [reply]
Ok the shader does this, noted!
The `sample_textures` does depend on `faces_per_pixel` and i did try it, before posting here, but this is along the z axis. I care about the x, y. If i have a cube for example, depending on the camera settings i might be close or far away. Being close and far away you get a different number of texels per pixel (what you would expect from minification and magnification). I dont care about the z axis.
